aten/src/ATen/native/native_functions.yaml

```diff
@@ -4914,6 +4914,11 @@
     SparseCsrCPU: _sparse_csr_sum_cpu
     SparseCsrCUDA: _sparse_csr_sum_cuda
 
+- func: _sparse_csr_prod.dim_dtype(Tensor self, int[1] dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
```

Comment:
Instead of a separate native function we could also consider scatter_reduce's approach, which takes a string as an input. Since this is a private function it is up to you, but add an argument seems cleaner for this specific case.

Overall however we should avoid adding any more reductions after this one before we deduplicated this code with that of segment_reduce

Reply:
> Overall however we should avoid adding any more reductions after this one before we deduplicated this code with that of segment_reduce

In addition, we should compare the performances of the different approaches. Although, it is not obvious to me how to use segment_reduce (that takes strided input) for CSR inputs. For the `dim=1` case, one could set up a loop that uses segment_reduce on CSR data, but it cannot be more performant than reduce_sparse_csr_dim1_cpu_template that takes full advantage of the CSR format properties. For the `dim=0` case, the complexity from torch.unique is the main bottleneck that, IIUC, cannot be avoided when using the segment_reduce approach either.

Reply:
Yes, it only works for one case. In particular I'm looking at [pytorch_sparse's code](https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/reduce.py#L46-L47) for guidance here. segment_csr should be supportable via [segment_reduce](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L11328-L11336) by passing lengths (as determined by the compressed row indices) and the values as data.